src/main/scala/com/scality/clueso/MetadataIngestionPipeline.scala

```diff
@@ -0,0 +1,187 @@
+import org.apache.spark.sql.streaming.{OutputMode, ProcessingTime}
+import org.apache.spark.sql.types.TimestampType
+import org.apache.spark.sql.{DataFrame, SparkSession}
+
+import scala.collection.mutable
+
+class EventMessageRewriter {
+  val mapper = new ObjectMapper()
+
+  def rewriteMsg(bytes: Array[Byte]): String = {
+    val payload = new String(bytes)
+
+    var rootNode = mapper.readTree(payload).asInstanceOf[ObjectNode]
+
+    val valueTxtNode = rootNode.path("value").asInstanceOf[TextNode]
+
+    val valueNode = mapper.readTree(valueTxtNode.asText()).asInstanceOf[ObjectNode]
+
+    val metadataFieldNames = mutable.ListBuffer[String]()
+
+    val it = valueNode.fieldNames()
+    while (it.hasNext) {
+      val fieldName = it.next()
+      if (fieldName.startsWith("x-amz-meta-")) {
+        metadataFieldNames += fieldName
+      } else if (fieldName.equals("location")) {
+        val locationArray = valueNode.path("location")
+
+        if (!locationArray.isInstanceOf[NullNode] && locationArray.asInstanceOf[ArrayNode].size() > 1) {
+          val _locationArray =  locationArray.asInstanceOf[ArrayNode]
+          (1 until locationArray.size()).foreach(i => _locationArray.remove(i))
+          valueNode.replace("location", _locationArray)
+        }
+      }
+    }
+
+    val userMd = if (rootNode.has("userMd")) {
+      valueNode.path("userMd").asInstanceOf[ObjectNode]
+    } else {
+      valueNode.putObject("userMd")
+    }
+
+    for (fieldName <- metadataFieldNames) {
+      userMd.set(fieldName, valueNode.path(fieldName).deepCopy().asInstanceOf[JsonNode])
+      valueNode.remove(fieldName)
+    }
+
+    valueNode.replace("userMd", userMd)
+    rootNode.replace("value", valueNode)
+
+    mapper.writeValueAsString(rootNode)
+  }
+}
+
+object EventMessageRewriterWrapper {
+  val deser = new EventMessageRewriter
+}
+
+object MetadataIngestionPipeline extends LazyLogging {
+
+  val msgRewriteFun = (bytes: Array[Byte]) =>
+    EventMessageRewriterWrapper.deser.rewriteMsg(bytes)
+
+  val maxOpIndexFun = (compactionRecordInterval:Long, value : String) => {
+    val recordNo = value.substring(0, 12).toLong
+
+    if (recordNo % compactionRecordInterval == 0) {
+      recordNo
+    } else {
+      recordNo + compactionRecordInterval - recordNo % compactionRecordInterval
+    }
+  }
+
+  import org.apache.spark.sql.functions.udf
+  val msg_rewrite = udf(msgRewriteFun)
+  val max_op_index = udf(maxOpIndexFun)
+
+  /**
+    * Applies projections and conditions to incoming data frame
+    * Ignores malformed json and garbage payloads
+    * Ignores events from specified bucket name
+    *
+    * @param bucketNameToFilterOut
+    * @param eventStream
+    * @return
+    */
+  def filterAndParseEvents(bucketNameToFilterOut : String, eventStream: DataFrame)(implicit spark : SparkSession, config: CluesoConfig) = {
+
+    var df = eventStream.select(
+      col("timestamp").cast(TimestampType).as("kafkaTimestamp"),
```

Comment:
why do we do this?

Reply:
The 'timestamp' field comes from Kafka. Kafka 0.10+ supports timestamps in every record. Now we should probably be using the last-modified, but this was the way we started. We can discuss later today